### DISCUSSION TITLE
How to add custom metadata to the Events generated by a Device Service ? ( Eg. Modbus Service )

### DISCUSSION THREAD

**Discussion by FabioBoarettiDosSantos** [edgexfoundry/community#43] [Q&A]
Please,

Can I have some guidance how to include more METADATA to the event.
I`m working with the IRELAND version at moment. MODBUS GOLANG

{"apiVersion":"v2","requestId":"20c74da2-849d-4955-a1d4-7eaffdf74152","event":{"apiVersion":"v2","id":"299f4034-815e-4a29-b611-e4fd6181b6f4","deviceName":"DGS-SOUST","profileName":"DGS-EXT","sourceName":"ReadyStatusBit","origin":1675107969704251870,"readings":[{"id":"bc3b339b-29b2-45d2-9109-2c95ff7cd79a","origin":1675107969704251870,"deviceName":"DGS-SOUST","resourceName":"ReadyStatusBit","profileName":"DGS-EXT","valueType":"String","value":"READY"}]}}

Thank You So Much [ ]

**Comment by cloudxxx8** [comment]
Event contains a `tags` field for adding more additional information.  You can add more information in the App Service level in the Ireland release.  Device Service didn't support it.
https://docs.edgexfoundry.org/2.3/microservices/general/#event-tagging

We are developing EdgeX 3.0, and you can add tags information into the Device metadata, and it can be added to the Event by Device Service.
https://github.com/edgexfoundry/go-mod-core-contracts/issues/767

**Reply by FabioBoarettiDosSantos** [reply]
@cloudxxx8 Thank you so much for your inputs and to be working on that feature! 
If you allow me a following question about it: in the deviceProfile there is a "tag" but I cannot see it been
pushed in the event or readings ( is it part of the edgex 3.0) . Is it disable / no-functional or I`m doing something wrong ? 

Thank you so much !

**Reply by FabioBoarettiDosSantos** [reply]
ADD: . Optionally I will use Edgex 2.2.

**Reply by lenny-goodell** [reply]
> ADD: . Optionally I will use Edgex 2.2.

Please use latest 2.3 which was released in Nov 2022

**Reply by cloudxxx8** [reply]
> @cloudxxx8 Thank you so much for your inputs and to be working on that feature! If you allow me a following question about it: in the deviceProfile there is a "tag" but I cannot see it been pushed in the event or readings ( is it part of the edgex 3.0) . Is it disable / no-functional or I`m doing something wrong ?
> 
> Thank you so much !

The data type of `tag` was wrong, and we didn't really use it, so we removed it in 3.0.  Please use `tags` in the future release.
Before 3.0, you need to implement your own logic in Device Service or App Service for the metadata you need.
